components/ai_chat/browser/ai_chat_feedback_api.cc

```diff
@@ -0,0 +1,138 @@
+
+#include <base/containers/flat_map.h>
+
+#include <utility>
+
+#include "base/json/json_writer.h"
+#include "base/no_destructor.h"
+#include "base/strings/strcat.h"
+#include "base/values.h"
+#include "brave/brave_domains/buildflags.h"
+#include "brave/components/brave_stats/browser/brave_stats_updater_util.h"
+#include "brave/components/l10n/common/locale_util.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "url/gurl.h"
+#include "url/url_constants.h"
+
+namespace ai_chat {
+namespace {
+
+constexpr char kFeedbackHostnamePart[] = "feedback";
+constexpr char kRatingPath[] = "1/ai/feedback/rating";
+constexpr char kFeedbackFormPath[] = "1/ai/feedback/form";
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("ai_chat", R"(
+      semantics {
+        sender: "AI Chat Feedback"
+        description:
+          "This is used to communicate with a service to record a rating"
+          "of an AI Chat message, and anonymous details about that message."
+        trigger:
+          "Triggered by user choosing a rating for a received AI message."
+        data:
+          "Positive or negative rating along with an extra details the user
+          provides, and the associated chat conversation messages."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+std::string CreateJSONRequestBody(base::ValueView node) {
+  std::string json;
+  base::JSONWriter::Write(node, &json);
+  return json;
+}
+
+base::Value::Dict ToRatingPayloadDict(const RatingPayload& rating_payload) {
+  base::Value::Dict dict;
+  dict.Set("id", rating_payload.id);
+  dict.Set("message", rating_payload.message);
+
+  return dict;
+}
+
+const GURL GetEndpointBaseUrl() {
+  auto* services_domain = BUILDFLAG(BRAVE_SERVICES_PRODUCTION_DOMAIN);
+
+  static base::NoDestructor<GURL> url{
+      base::StrCat({url::kHttpsScheme, url::kStandardSchemeSeparator,
+                    kFeedbackHostnamePart, ".", services_domain})};
+  return *url;
+}
+
+}  // namespace
+
+AIChatFeedbackAPI::AIChatFeedbackAPI(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    const std::string& channel_name)
+    : api_request_helper_(GetNetworkTrafficAnnotationTag(), url_loader_factory),
+      channel_name_(channel_name) {}
+
+AIChatFeedbackAPI::~AIChatFeedbackAPI() = default;
+
+void AIChatFeedbackAPI::SendRating(
+    bool is_liked,
+    RatingPayload rating_payload,
+    const std::string& model_name,
+    api_request_helper::APIRequestHelper::ResultCallback on_complete_callback) {
+  base::Value::Dict dict;
+  base::Value::List chat;
+
+  chat.Append(ToRatingPayloadDict(rating_payload));
```

Comment:
We're only including the message being rated, but the spec is that the entire conversation, up to and including the rated message, is sent. That's why this is an array.

Reply:
sending the history array up to the rated message now